maxfw/utils/image_functions.py

```diff
@@ -0,0 +1,443 @@
+        if pic.ndim not in {2, 3}:
+            raise ValueError('pic should be 2/3 dimensional. Got {} dimensions.'.format(pic.ndim))
+
+        elif pic.ndim == 2:
+            # if 2D image, add channel dimension (HWC)
+            pic = np.expand_dims(pic, 2)
+
+    elif isinstance(pic, (bytes, bytearray)):
+        try:
+            # verify that the object can be loaded into memory
+            pic = np.array(Image.open(io.BytesIO(pic)))
+        except Exception:
+            raise TypeError('The input bytes object is not suitable for the Pillow library. Check the input again.')
+
+    npimg = pic
+    if not isinstance(npimg, np.ndarray):
+        raise TypeError('Input pic must be a bytes object or NumPy ndarray, ' +
+                        'not {}'.format(type(npimg)))
+
+    if npimg.shape[2] == 1:
+        expected_mode = None
+        npimg = npimg[:, :, 0]
+        if npimg.dtype == np.uint8:
+            expected_mode = 'L'
+        elif npimg.dtype == np.int16:
+            expected_mode = 'I;16'
+        elif npimg.dtype == np.int32:
+            expected_mode = 'I'
+        elif npimg.dtype == np.float32:
+            expected_mode = 'F'
+        if mode is not None and mode != expected_mode:
+            raise ValueError("Incorrect mode ({}) supplied for input type {}. Should be {}"
+                             .format(mode, np.dtype, expected_mode))
+        mode = expected_mode
+
+    elif npimg.shape[2] == 2:
+        permitted_2_channel_modes = ['LA']
+        if mode is not None and mode not in permitted_2_channel_modes:
+            raise ValueError("Only modes {} are supported for 2D inputs".format(permitted_2_channel_modes))
+
+        if mode is None and npimg.dtype == np.uint8:
+            mode = 'LA'
+
+    elif npimg.shape[2] == 4:
+        permitted_4_channel_modes = ['RGBA', 'CMYK', 'RGBX']
+        if mode is not None and mode not in permitted_4_channel_modes:
+            raise ValueError("Only modes {} are supported for 4D inputs".format(permitted_4_channel_modes))
+
+        if mode is None and npimg.dtype == np.uint8:
+            mode = 'RGBA'
+    else:
+        permitted_3_channel_modes = ['RGB', 'YCbCr', 'HSV']
+        if mode is not None and mode not in permitted_3_channel_modes:
+            raise ValueError("Only modes {} are supported for 3D inputs".format(permitted_3_channel_modes))
+        if mode is None and npimg.dtype == np.uint8:
+            mode = 'RGB'
+
+    if mode is None:
+        raise TypeError('Input type {} is not supported'.format(npimg.dtype))
+
+    # Verify that the target mode exists
+    assert target_mode in [1, 'L', 'P', 'RGB', 'RGBA', 'CMYK', 'YCbCr', 'LAB', 'HSV', 'I', 'F', 'RGBX', 'RGBBa']
+
+    return Image.fromarray(npimg, mode=mode).convert(target_mode)
+
+
+def normalize(img):
+    if type(img) is not np.ndarray:
+        img = np.array(img)
+    img = img / (np.max(img) - np.min(img))
+    return Image.fromarray(img.astype('uint8'))
+
+
+def standardize(img):
```

Comment:
Just a note that we will need to support channel-wise values for `mean` and `std` (e.g. we will need to be able to replace the [Keras model image pre-processing](https://github.com/keras-team/keras-applications/blob/master/keras_applications/imagenet_utils.py)).

Also note, typically this function is going to standardize based on a pre-computed `mean` and `std` (i.e. over the entire training set). Doing this per image seems unusual to me, but perhaps there are cases where that happens. We can certainly have that as an option and perhaps the default behavior if there's no `mean` and `std` passed in...

This needs to be documented pretty clearly.

Reply:
Good point, need to rework this @splovyt.